spec.bs

```diff
@@ -512,42 +512,19 @@ To <dfn>validate and convert auction ad config</dfn> given an {{AuctionAdConfig}
-      1. Let |pendingException| be null.
-      1. If |result| is not a [=record=] mapping from {{USVString}} to {{any}}, set |pendingException| to
-        a {{TypeError}} and jump to the step labeled
-        <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
-      1. Set |auctionConfig|'s [=auction config/per buyer signals=] to a new [=ordered map=] whose
-        [=map/keys=] are [=origins=] and whose [=map/values=] are [=strings=].
-      1. [=map/For each=] |key| → |value| of |result|:
-        1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
-          set |pendingException| to a {{TypeError}} and jump to the step labeled
-          <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
-        1. Let |signalsString| be the result of [=serializing a JavaScript value to a JSON string=], given
-          |value|. If this [=exception/throws=] an exception |e|, set |pendingException| to |e| and jump
-          to the step labeled <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
-        1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to
-          |signalsString|.
-      1. <i id=validate-per-buyer-signals-finish>Finish</i>:
-        1. If |pendingException| is not null:
-          1. Set |auctionConfig|'s [=auction config/per buyer signals=] to failure.
-          1. Decrement |auctionConfig|'s [=auction config/pending promise count=].
-          1. [=exception/Throw=] |pendingException|.
-        1. Otherwise, decrement |auctionConfig|'s [=auction config/pending promise count=].
-    1. [=Upon rejection=] of |auctionConfig|'s [=auction config/per buyer signals=]:
+  1. Set |auctionConfig|'s [=auction config/per buyer signals=] to
+    |config|["{{AuctionAdConfig/perBuyerSignals}}"].
+  1. Increment |auctionConfig|'s [=auction config/pending promise count=].
+  1. [=Upon fulfillment=] of |auctionConfig|'s [=auction config/per buyer signals=] with |result|:
+    1. Execute [=validate and convert per buyer signals=] with |auctionConfig| and |result|.
+    1. If the previous step [=exception/throws=] an exception |e|:
       1. Set |auctionConfig|'s [=auction config/per buyer signals=] to failure.
       1. Decrement |auctionConfig|'s [=auction config/pending promise count=].
-  1. [=map/For each=] |key| → |value| of |config|["{{AuctionAdConfig/perBuyerSignals}}"]:
-    1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
-      [=exception/throw=] a {{TypeError}}.
-    1. Let |signalsString| be the result of [=serializing a JavaScript value to a JSON string=], given
-      |value|.
-    1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to
-      |signalsString|.
+      1. [=exception/Throw=] |e|.
+    1. Otherwise, decrement |auctionConfig|'s [=auction config/pending promise count=].
+  1. [=Upon rejection=] of |auctionConfig|'s [=auction config/per buyer signals=]:

```

Comment:
Here, I believe you want to react to any exceptions thrown by the "Upon fulfillment" steps, including the type error if the Promise didn't hold a record, but since you're using "upon rejection" on the original promise, you won't see them. Instead, you'll need to catch the result of "upon fulfillment" with some awkward wording:

```
1. Let |perBuyerSignalsResult| be the result of running the following steps
    [=upon fulfillment=] of |config|["{{AuctionAdConfig/perBuyerSignals}}"] with |result|:
```

(Or staying with `|auctionConfig|'s [=auction config/per buyer signals=]` if you do need it to have an intermediate Promise value.)

Then this step becomes
```suggestion
  1. [=Upon rejection=] of |perBuyerSignalsResult|:
```

Reply:
Sorry, could you elaborate? What happens in the spec land if Promise<T> resolves to something other than a T? Does it accept/reject the promise? Do some secret other third thing? If it rejects, then that's perfectly fine, it will do the proper thing anyway. 

(I am also having a really hard time parsing your example).

(Chrome's WebIDL doesn't actually check that Promise<T> resolves to a T).

Reply:
Oh, I see, it's doing promise chaining. Huh, but then, I mostly don't need to explicitly handle the failure myself, do I? Won't it end up in the catch/on reject anyway? Anyway, tried in a way that's actually comprehensible to me, and back to inlined conversion since it got simple enough.  It's a bit awkward, but it gets into why I didn't understand your suggestion: it sounded me like you were binding |perBuyerSignalsResult| to the result of /steps/ rather than to the promise made by [=upon fullfillment=].

(Also, is there anything that actually uses it out there? The only use I found in Chrome via crude scanning of codesearch output for taking it as input is PromiseRejectionEventInit, which takes... Promise<any>; at any rate the impl doesn't do any <T> checking on promises anyway).